src/table-procedure/src/drop.rs

```diff
@@ -0,0 +1,295 @@
+        Ok(json)
+    }
+
+    fn lock_key(&self) -> LockKey {
+        // We lock the whole table.
+        let table_name = self.data.table_ref().to_string();
+        LockKey::single(table_name)
+    }
+}
+
+impl DropTableProcedure {
+    const TYPE_NAME: &str = "table-procedure::DropTableProcedure";
+
+    /// Returns a new [DropTableProcedure].
+    pub fn new(
+        request: DropTableRequest,
+        catalog_manager: CatalogManagerRef,
+        engine_procedure: TableEngineProcedureRef,
+    ) -> DropTableProcedure {
+        DropTableProcedure {
+            data: DropTableData {
+                state: DropTableState::Prepare,
+                request,
+                subprocedure_id: None,
+            },
+            catalog_manager,
+            engine_procedure,
+        }
+    }
+
+    /// Register the loader of this procedure to the `procedure_manager`.
+    ///
+    /// # Panics
+    /// Panics on error.
+    pub fn register_loader(
+        catalog_manager: CatalogManagerRef,
+        engine_procedure: TableEngineProcedureRef,
+        procedure_manager: &dyn ProcedureManager,
+    ) {
+        procedure_manager
+            .register_loader(
+                Self::TYPE_NAME,
+                Box::new(move |data| {
+                    Self::from_json(data, catalog_manager.clone(), engine_procedure.clone())
+                        .map(|p| Box::new(p) as _)
+                }),
+            )
+            .unwrap()
+    }
+
+    /// Recover the procedure from json.
+    fn from_json(
+        json: &str,
+        catalog_manager: CatalogManagerRef,
+        engine_procedure: TableEngineProcedureRef,
+    ) -> Result<Self> {
+        let data: DropTableData = serde_json::from_str(json).context(DeserializeProcedureSnafu)?;
+
+        Ok(DropTableProcedure {
+            data,
+            catalog_manager,
+            engine_procedure,
+        })
+    }
+
+    async fn on_prepare(&mut self) -> Result<Status> {
+        let request = &self.data.request;
+        // Ensure the table exists.
+        self.catalog_manager
+            .table(
+                &request.catalog_name,
+                &request.schema_name,
+                &request.table_name,
+            )
+            .await
+            .context(AccessCatalogSnafu)?
+            .context(TableNotFoundSnafu {
+                name: &request.table_name,
+            })?;
+
+        self.data.state = DropTableState::RemoveFromCatalog;
+
+        Ok(Status::executing(true))
+    }
+
+    async fn on_remove_from_catalog(&mut self) -> Result<Status> {
+        let deregister_table_req = DeregisterTableRequest {
+            catalog: self.data.request.catalog_name.clone(),
+            schema: self.data.request.schema_name.clone(),
+            table_name: self.data.request.table_name.clone(),
+        };
+        self.catalog_manager
+            .deregister_table(deregister_table_req)
+            .await
+            .map_err(Error::from_error_ext)?;
```

Comment:
If this procedure were to be executed again (to keep idempotence), shall we check the returned errors of `deregister_table`, that if `deregister_table` fails with "table not found", procedure can continue?

Reply:
Oops, looks like the behavior of `deregister_table()` is a little different from the document.
https://github.com/GreptimeTeam/greptimedb/blob/e4cd08c750cb830e21b7ecf2ee60ff331282244f/src/catalog/src/lib.rs#L96-L98

The implementation actually returns an error
https://github.com/GreptimeTeam/greptimedb/blob/e4cd08c750cb830e21b7ecf2ee60ff331282244f/src/catalog/src/local/manager.rs#L468-L474

So I prepare to check the table's existence first.